<a href="https://colab.research.google.com/github/singithi03/Data-Analytics_s5214680/blob/Lab---5/BOW(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 02-BOW

Now that we master the preprocessing, let's make our first Bag Of Words (BOW).

We will reuse our dataset of Coldplay songs to make a BOW.

As usual, the first step is to import some libraries. So import *nltk* as well as all the libraries you will need.

In [1]:
# Import NLTK and all the needed libraries
import nltk
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

Load now the dataset in *coldplay.csv* using pandas.

In [3]:
# TODO: Load the dataset in coldplay.csv
df = pd.read_csv('coldplay.csv')
df

Artist                           Song  \
0    Coldplay                 Another's Arms   
1    Coldplay                Bigger Stronger   
2    Coldplay                       Daylight   
3    Coldplay                       Everglow   
4    Coldplay  Every Teardrop Is A Waterfall   
..        ...                            ...   
115  Coldplay           Hymn For The Weekend   
116  Coldplay                    In My Place   
117  Coldplay                            Ink   
118  Coldplay              Ladder To The Sun   
119  Coldplay                           Lost   

                                                  Link  \
0              /c/coldplay/anothers+arms_21079526.html   
1            /c/coldplay/bigger+stronger_20032648.html   
2                   /c/coldplay/daylight_20032625.html   
3                   /c/coldplay/everglow_21104546.html   
4    /c/coldplay/every+teardrop+is+a+waterfall_2091...   
..                                                 ...   
115     /c/coldplay/hymn+for+the+weekend_21104544.html   
116              /c/coldplay/in+my+place_20032629.html   
117                      /c/coldplay/ink_21082518.html   
118        /c/coldplay/ladder+to+the+sun_20232934.html   
119                     /c/coldplay/lost_20743853.html   

                                                Lyrics  
0    Late night watching tv  \nUsed to be you here ...  
1    I want to be bigger stronger drive a faster ca...  
2    To my surprise, and my delight  \nI saw sunris...  
3    Oh, they say people come  \nThey say people go...  
4    I turn the music up, I got my records on  \nI ...  
..                                                 ...  
115  Oh, angel sent from up above  \nYou know you m...  
116  In my place, in my place  \nWere lines that I ...  
117  Got a tattoo that says "2gether thru life"  \n...  
118  From the very start  \nIt came apart  \nIt bro...  
119  Just because I'm losing  \nDoesn't mean I'm lo...  

[120 rows x 4 columns]

You already know this dataset, but you can check it again if you want to refresh your memory.

In [4]:
# TODO: Explore the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Artist  120 non-null    object
 1   Song    120 non-null    object
 2   Link    120 non-null    object
 3   Lyrics  120 non-null    object
dtypes: object(4)
memory usage: 3.9+ KB


Now using the *CountVectorizer* of scikit-learn, make a BOW of all the lyrics of Coldplay, and print the result.

In [8]:
# TODO: Compute a BOW
vectorize = CountVectorizer(stop_words='english')
BOW = vectorize.fit_transform(df['Lyrics']).toarray()
BOW.shape

(120, 1569)

Now that we have the BOW matrix, we would like to have a new dataframe having the BOW for each song, and as columns the corresponding words (just as we did in the lecture at the end).

So that at the end we would end up with a dataframe containing something like the following (120 raws for 120 songs, and as many columns as words):

| | ah | adventure | ... | yeah 
|---|---|---|---|---| 
| 0 | 0 | 1 | ... | 4 |
| 1 | 8 | 0 | ... | 2 |
|...|...|...|...|...|
| 119 | 5 | 0 | ... | 8 |

In [9]:
# TODO: Create a new dataframe containing the BOW outputs and the corresponding words as columns. And print it
tokens = vectorize.get_feature_names()
bow_df = pd.DataFrame(data=BOW, columns=tokens)
bow_df

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


10  2000  2gether  76543  aaaaaah  aaaaah  aaaah  achin  adventure  \
0     0     0        0      0        0       0      0      0          0   
1     0     0        0      0        0       0      0      0          0   
2     0     0        0      0        0       0      0      0          0   
3     0     0        0      0        0       0      0      0          0   
4     0     0        0      0        0       0      0      0          0   
..   ..   ...      ...    ...      ...     ...    ...    ...        ...   
115   0     0        0      0        0       0      0      0          0   
116   0     0        0      0        0       0      0      0          0   
117   0     0        1      0        0       0      0      0          0   
118   0     0        0      0        0       0      0      0          0   
119   0     0        0      0        0       0      0      0          0   

     advice  ...  x2  x7  ya  yeah  years  yellow  yes  yesterday  young  \
0         0  ...   0   0   0     0      0       0    0          0      0   
1         0  ...   0   0   0     0      0       0    0          0      0   
2         0  ...   0   0   0     2      0       0    0          0      0   
3         0  ...   0   0   0     2      0       0    0          0      0   
4         0  ...   0   0   0     0      0       0    0          0      0   
..      ...  ...  ..  ..  ..   ...    ...     ...  ...        ...    ...   
115       0  ...   0   0   0     0      0       0    0          0      0   
116       0  ...   0   0   0    11      0       0    0          0      0   
117       0  ...   0   0   0     3      0       0    0          0      0   
118       0  ...   0   0   0     0      0       0    0          0      0   
119       0  ...   0   0   0     0      0       0    0          0      0   

     yuletide  
0           0  
1           0  
2           0  
3           0  
4           0  
..        ...  
115         0  
116         0  
117         0  
118         0  
119         0  

[120 rows x 1569 columns]

Well as you see we're still having some issue, we have some tokens that are not words, like '10' or '2000'.

To get rid of that, we could use directly regular expressions within the function. Another solution would be to make preprocessing before using the function *CountVectorizer*.

For the moment, we won't pay attention to this issue. But if you are curious and have time, you can find on google how to remove those words using the *CountVectorizer*.

Now we would like to see what are the most used words by Coldplay.

In [10]:
sum_bow = bow_df.sum()
sum_bow.idxmax()

'oh'

So what is the most used word? Are you surprised?

Now make a sort in order to show the 10 most used.

In [14]:
# TODO: print the 10 most used word by Coldplay
sum_bow.sort_values(ascending=False)[:10]

oh      334
don     190
know    137
just    136
ll      132
come    126
yeah    111
love     95
ooh      95
want     86
dtype: int64

Here it is! You know the Coldplay lyrics more than the singers now!